File: example_programs/assembler.py
import re
import pathlib
import argparse


def check_mov(args):
    r = re.match(r"([abcd]) ([abcd])", args)
    if r is not None:
        return r.group(1) != r.group(2)


def check_load(args):
    r = re.match(r"ram\[[abcd]\] [abcd]", args)
    if r is not None:
        return True
    r = re.match(r"ram\[[0-9]+\] [abcd]", args)
    if r is not None:
        return True
    r = re.match(r"rom [abcd] [0-9]+", args)
    return r is not None


def check_save(args):
    r = re.match(r"[abcd] ram", args)
    if r is not None:
        return True
    r = re.match(r"[abcd] ram\[[abcd]\]", args)
    if r is not None:
        return True
    r = re.match(r"[abcd] ram\[[0-9]+\]", args)
    if r is not None:
        return True
    r = re.match(r"[abcd] mar", args)
    return r is not None


operations = {
    "nop": lambda x: x == "",
    "mov": check_mov,
    "cmp": re.compile(r"([abcd]) ([abcd])").match,
    "jmp": re.compile(r"(\.?(<=|<|=|>|>=) [abcd])|(.+)").match,
    "jmpr": re.compile(r"(\.?(<=|<|=|>|>=) [abcd])|(.+)").match,
    "opp": re.compile(r"").match,
    "load": check_load,
    "save": check_save,
    "in": re.compile(r"[abcd]").match,
    "out": re.compile(r"[abcd]").match,
    "pause": lambda x: x == "",
}

translation = {
    "nop": 0x0,
    "mov a b": 0x1,
    "mov a c": 0x2,
    "mov a d": 0x3,
    "mov b a": 0x4,
    "mov b c": 0x5,
    "mov b d": 0x6,
    "mov c a": 0x7,
    "mov c b": 0x8,
    "mov c d": 0x9,
    "mov d a": 0xA,
    "mov d b": 0xB,
    "mov d c": 0xC,
    "cmp a a": 0x10,
    "cmp a b": 0x11,
    "cmp a c": 0x12,
    "cmp a d": 0x13,
    "cmp b a": 0x14,
    "cmp b b": 0x15,
    "cmp b c": 0x16,
    "cmp b d": 0x17,
    "cmp c a": 0x18,
    "cmp c b": 0x19,
    "cmp c c": 0x1A,
    "cmp c d": 0x1B,
    "cmp d a": 0x1C,
    "cmp d b": 0x1D,
    "cmp d c": 0x1E,
    "cmp d d": 0x1F,
    "jmp {label}": 0x20,
    "jmp = {label}": 0x21,
    "jmp != {label}": 0x22,
    "jmp < {label}": 0x23,
    "jmp <= {label}": 0x24,
    "jmp > {label}": 0x25,
    "jmp >= {label}": 0x26,
    "jmp .< {label}": 0x27,
    "jmp .<= {label}": 0x28,
    "jmp .> {label}": 0x29,
    "jmp .>= {label}": 0x2A,
    "jmpr {number}": 0x30,
    "jmpr = {number}": 0x31,
    "jmpr != {number}": 0x32,
    "jmpr < {number}": 0x33,
    "jmpr <= {number}": 0x34,
    "jmpr > {number}": 0x35,
    "jmpr >= {number}": 0x36,
    "jmpr .< {number}": 0x37,
    "jmpr .<= {number}": 0x38,
    "jmpr .> {number}": 0x39,
    "jmpr .>= {number}": 0x3A,
    "opp clr": 0x40,
    "opp cmp off": 0x41,
    "opp cmp on": 0x42,
    "opp sign off": 0x43,
    "opp sign on": 0x44,
    "opp 0": 0x45,
    "opp 1": 0x46,
    "opp -1": 0x47,
    "opp a": 0x48,
    "opp b": 0x49,
    "opp c": 0x4A,
    "opp d": 0x4B,
    "opp ~a": 0x4C,
    "opp ~b": 0x4D,
    "opp ~c": 0x4E,
    "opp ~d": 0x4F,
    "opp -a": 0x50,
    "opp -b": 0x51,
    "opp -c": 0x52,
    "opp -d": 0x53,
    "opp a+1": 0x54,
    "opp b+1": 0x55,
    "opp c+1": 0x56,
    "opp d+1": 0x57,
    "opp a-1": 0x58,
    "opp b-1": 0x59,
    "opp c-1": 0x5A,
    "opp d-1": 0x5B,
    "opp a+b": 0x5C,
    "opp a+c": 0x5D,
    "opp a+d": 0x5E,
    "opp b+a": 0x5F,
    "opp b+c": 0x60,
    "opp b+d": 0x61,
    "opp c+a": 0x62,
    "opp c+b": 0x63,
    "opp c+d": 0x64,
    "opp d+a": 0x65,
    "opp d+b": 0x66,
    "opp d+c": 0x67,
    "opp a-b": 0x68,
    "opp a-c": 0x69,
    "opp a-d": 0x6A,
    "opp b-a": 0x6B,
    "opp b-c": 0x6C,
    "opp b-d": 0x6D,
    "opp c-a": 0x6E,
    "opp c-b": 0x6F,
    "opp c-d": 0x70,
    "opp d-a": 0x71,
    "opp d-b": 0x72,
    "opp d-c": 0x73,
    "opp a*a": 0x74,
    "opp a*b": 0x75,
    "opp a*c": 0x76,
    "opp a*d": 0x77,
    "opp b*a": 0x78,
    "opp b*b": 0x79,
    "opp b*c": 0x7A,
    "opp b*d": 0x7B,
    "opp c*a": 0x7C,
    "opp c*b": 0x7D,
    "opp c*c": 0x7E,
    "opp c*d": 0x7F,
    "opp d*a": 0x80,
    "opp d*b": 0x81,
    "opp d*c": 0x82,
    "opp d*d": 0x83,
    "opp a.*a": 0x84,
    "opp a.*b": 0x85,
    "opp a.*c": 0x8D,
    "opp c.*c": 0x8E,
    "opp c.*d": 0x8F,
    "opp d.*a": 0x90,
    "opp d.*b": 0x91,
    "opp d.*c": 0x92,
    "opp d.*d": 0x93,
    "opp a/b": 0x94,
    "opp a/c": 0x95,
    "opp a/d": 0x96,
    "opp b/a": 0x97,
    "opp b/c": 0x98,
    "opp b/d": 0x99,
    "opp c/a": 0x9A,
    "opp c/b": 0x9B,
    "opp c/d": 0x9C,
    "opp d/a": 0x9D,
    "opp d/b": 0x9E,
    "opp d/c": 0x9F,
    "opp a&b": 0xA0,
    "opp a&c": 0xA1,
    "opp a&d": 0xA2,
    "opp b&c": 0xA3,
    "opp b&d": 0xA4,
    "opp c&d": 0xA5,
    "opp a|b": 0xA6,
    "opp a|c": 0xA7,
    "opp a|d": 0xA8,
    "opp b|c": 0xA9,
    "opp b|d": 0xAA,
    "opp c|d": 0xAB,
    "load ram[a] a": 0xB0,
    "load ram[a] b": 0xB1,
    "load ram[a] c": 0xB2,
    "load ram[a] d": 0xB3,
    "load ram[b] a": 0xB4,
    "load ram[b] b": 0xB5,
    "load ram[b] c": 0xB6,
    "load ram[b] d": 0xB7,
    "load ram[c] a": 0xB8,
    "load ram[c] b": 0xB9,
    "load ram[c] c": 0xBA,
    "load ram[c] d": 0xBB,
    "load ram[d] a": 0xBC,
    "load ram[d] b": 0xBD,
    "load ram[d] c": 0xBE,
    "load ram[d] d": 0xBF,
    "load rom a {number}": 0xC0,
    "load rom b {number}": 0xC1,
    "load rom c {number}": 0xC2,
    "load rom d {number}": 0xC3,
    "load ram[{number}] a": 0xC4,
    "load ram[{number}] b": 0xC5,
    "load ram[{number}] c": 0xC6,
    "load ram[{number}] d": 0xC7,
    "set a rampage": 0xC8,
    "set b rampage": 0xC9,
    "set c rampage": 0xCA,
    "set d rampage": 0xCB,
    "save a mar": 0xD0,
    "save b mar": 0xD1,
    "save c mar": 0xD2,
    "save d mar": 0xD3,
    "save a ram[current]": 0xD4,
    "save b ram[current]": 0xD5,
    "save c ram[current]": 0xD6,
    "save d ram[current]": 0xD7,
    "save a ram[a]": 0xD8,
    "save b ram[c]": 0xD9,
    "save c ram[d]": 0xDA,
    "save d ram[d]": 0xDB,
    "save a ram[{number}]": 0xDC,
    "save b ram[{number}]": 0xDD,
    "save c ram[{number}]": 0xDE,
    "save d ram[{number}]": 0xDF,
    "in a": 0xE0,
    "in b": 0xE1,
    "in c": 0xE2,
    "in d": 0xE3,
    "out a": 0xF0,
    "out b": 0xF1,
    "out c": 0xF2,
    "out d": 0xF3,
    "pause": 0xFF,
}

translation_stage_two = list(
    filter(lambda x: "{label}" in x or "{number}" in x, translation)
)


def opp_to_hex(line):
    global offset

    if line in translation:
        offset += 1
        return [translation[line]]

    for instruction in translation_stage_two:
        match_whole_ins = "^" + re.escape(instruction) + "$"

        ins_temp = match_whole_ins.replace("\{label\}", "([^ ]+)")
        match = re.match(ins_temp, line)
        if match is not None:
            # Instructions with labels are 3 bytes
            offset += 3
            return [translation[instruction], *match.groups()]

        ins_temp = match_whole_ins.replace(
            "\{number\}", "(?:(0x([0-9a-fA-F]+))|(0b([01]+))|([0-9]+))"
        )
        match = re.match(ins_temp, line)
        if match is not None:
            # Instructions with numbers are 2 bytes
            offset += 2

            # TODO: Only supports one number per instruction
            if match.group(1):
                number = int(match.group(2), 16)
            elif match.group(3):
                number = int(match.group(4), 2)
            elif match.group(5):
                number = int(match.group(5))

            if number > 0xFF:
                print(line)
                print("Number larger than can fit in register")
                return None

            return [translation[instruction], number]
    return None


def parse(input_file, output_file):
    final = []
    labels = {}
    global offset
    offset = 0

    for ln, line in enumerate(input_file):
        line = re.sub(r"//.*", r"", line)
        line = line.strip()
        if not line:
            continue

        variables = line.split()
        opp = variables[0]
        opp_args = " ".join(variables[1:])

        label_match = re.match(":(.+)", opp)
        if label_match is not None:
            # print(line, offset)
            label_match = label_match.group(1)
            if label_match not in labels:
                labels[label_match] = offset
            else:
                print(line)
                print("Line %d duplicate label detected" % (ln + 1))
        elif opp in operations:
            if not operations[opp](opp_args):
                print(line)
                print("Line %d is not valid" % (ln + 1))
                return

            hex_op = opp_to_hex(line)
            if hex_op is None:
                print(line)
                print("Line %d couldn't find translation for instruction" % (ln + 1))
                return
            final.extend(hex_op)
        else:
            print(line)
            print("Line %d couldn't find matching instruction" % (ln + 1))
            return

    # for x in final:
    #    if isinstance(x, int):
    #        print(hex(int(x)), end=' ')
    #    else:
    #        print(x, end=' ')
    # print()

    # print(final)
    file_output = []
    for ins in final:
        if isinstance(ins, str):
            if ins not in labels:
                print("Label %s has not been defined" % (ins))
                return

            ins = labels[ins]
            file_output.append("%02x" % (ins >> 4 * 2))
            file_output.append("%02x" % (ins & 0xFF))
        else:
            file_output.append("%02x" % ins)
    # print(file_output)

    if output_file is None:
        output_file = open(pathlib.Path(input_file.name).stem + ".o", "w")

    output_file.write("v2.0 raw\n")
    output_file.write(" ".join(file_output))
    output_file.write("\n")
    output_file.close()

    print("Successfully compiled")
    print("File written to %s" % output_file.name)


parser = argparse.ArgumentParser(description="Process some integers.")
parser.add_argument(
    "input",
    type=argparse.FileType("r"),
    help="The assembly file to compile to machine level code",
)
parser.add_argument(
    "--output",
    "-o",
    type=argparse.FileType("w"),
    help="The machine level filename to write",
)
import sys

if len(sys.argv) > 1:
    args = parser.parse_args()
    input_file = args.input
    output_file = args.output
else:
    import tkinter as tk
    from tkinter import filedialog

    root = tk.Tk()
    root.withdraw()

    filename = filedialog.askopenfilename()
    input_file = open(filename, "r")
    output_file = None

parse(input_file, output_file)

File: compiler/assembler.nim
import os
import nre
import tables
import streams
import options
import parseopt
import strutils
import sequtils
import strformat

import cu_flags

type HexOrLabelKind = enum
  HEX
  LABEL

type HexOrLabel = object
  case kind: HexOrLabelKind
  of HEX:
    hex: int
  of LABEL:
    label: string

# Define checks
proc checkMov(args: string): bool =
  let r = args.match(re"([abcd]) ([abcd])")
  r.isSome() and r.get().captures()[0] != r.get().captures()[1]

proc checkLoad(args: string): bool =
  if args.match(re"ram\[[abcd]\] [abcd]").isSome():
    return true
  if args.match(re"ram\[[0-9]+\] [abcd]").isSome():
    return true
  args.match(re"rom [abcd] [0-9]+").isSome()

proc checkSave(args: string): bool =
  if args.match(re"[abcd] ram").isSome():
    return true
  if args.match(re"[abcd] ram\[[abcd]\]").isSome():
    return true
  if args.match(re"[abcd] ram\[[0-9]+\]").isSome():
    return true
  args.match(re"[abcd] mar").isSome()

# Define operations
var operations = {
  "nop": proc(x: string): bool =
    x == "",
  "mov": checkMov,
  "cmp": proc(x: string): bool =
    x.match(re"([abcd]) ([abcd]|0|1|-1|255)").isSome(),
  "jmp": proc(x: string): bool =
    x.match(re"(\.?(<=|<|=|>|>=) [abcd])|(.+)").isSome(),
  "jmpr": proc(x: string): bool =
    x.match(re"(\.?(<=|<|=|>|>=) [abcd])|(.+)").isSome(),
  "opp": proc(x: string): bool =
    x.match(re"").isSome(),
  "load": checkLoad,
  "save": checkSave,
  "in": proc(x: string): bool =
    x.match(re"[abcd]").isSome(),
  "out": proc(x: string): bool =
    x.match(re"[abcd]|[0-9]+|ram\[[0-9]+\]|ram\[[abcd]\]").isSome(),
  "pause": proc(x: string): bool =
    x == "",
}.toTable

let translation = cu_flags.CU_FLAGS
var translationKeys: seq[string] = @[]
for k in translation.keys:
  translationKeys.add(k)
var translationStageTwo =
  filterIt(translationKeys, it.contains(re"\{label\}") or it.contains(re"\{number\}"))

proc matchLabelInstruction(line: string, instruction: string): Option[seq[HexOrLabel]] =
  var matchWholeIns = fmt"^{escapeRe(instruction)}$".replace("\\{label\\}", "([^ ]+)")
  let match = line.match(re(matchWholeIns))
  if match.isSome():
    return some(
      @[
        HexOrLabel(kind: HexOrLabelKind.HEX, hex: translation[instruction]),
        HexOrLabel(kind: HexOrLabelKind.LABEL, label: match.get().captures()[0]),
      ]
    )
  none(seq[HexOrLabel])

proc matchNumberInstruction(
    line: string, instruction: string
): Option[seq[HexOrLabel]] =
  var matchWholeIns = fmt"^{escapeRe(instruction)}$".replace(
    "\\{number\\}", "((0x[0-9a-fA-F]+)|(0b[01]+)|(0o[0-9]+)|([0-9]+))"
  )
  let match = line.match(re(matchWholeIns))
  if match.isSome():
    let captures = toSeq(match.get().captures())
    let num = (
      if (captures[1].isSome()):
        parseHexInt(captures[1].get())
      elif (captures[2].isSome()):
        parseBinInt(captures[2].get())
      elif (captures[3].isSome()):
        parseOctInt(captures[3].get())
      else:
        parseInt(captures[4].get())
    )

    if num > 0xff:
      echo "Number larger than can fit in register"
      return none(seq[HexOrLabel])
    return some(
      @[
        HexOrLabel(kind: HexOrLabelKind.HEX, hex: translation[instruction]),
        HexOrLabel(kind: HexOrLabelKind.HEX, hex: num),
      ]
    )
  none(seq[HexOrLabel])

proc oppToHex(line: string, offset: var int): seq[HexOrLabel] =
  if translation.contains(line):
    offset.inc()
    return @[HexOrLabel(kind: HexOrLabelKind.HEX, hex: translation[line])]

  for instruction in translationStageTwo:
    let r = (
      if instruction.contains("{label}"):
        matchLabelInstruction(line, instruction)
      else:
        matchNumberInstruction(line, instruction)
    )

    if r.isSome():
      offset += (if instruction.contains("{label}"): 3 else: 2)
      return r.get()
  @[]

proc handleLabels(line: string, labels: var Table[string, int], offset: var int): bool =
  let labelMatch = line.match(re":(.+)")
  if labelMatch.isSome():
    let label = labelMatch.get().captures()[0].strip()
    if labels.contains(label):
      echo fmt"Duplicate label detected: {label}"
      return false
    labels[label] = offset
    return true
  return false

proc translateInstructions(
    line: string, offset: var int, final: var seq[HexOrLabel]
): bool =
  let variables = line.split()
  let opp = variables[0]
  let oppArgs = join(variables[1 ..^ 1], " ")

  if opp in operations:
    if not operations[opp](oppArgs):
      echo fmt"Invalid operation: {line}"
      return false
    let hexOp = oppToHex(line, offset)
    if hexOp.len == 0:
      echo fmt"Could not translate instruction: {line}"
      return false
    final.add(hexOp)
    return true
  else:
    echo fmt"Unrecognized instruction: {line}"
    return false

proc writeOutput(
    final: seq[HexOrLabel], labels: Table[string, int], outputFile: string
) =
  let outputFileStream = newFileStream(outputFile, fmWrite)

  var fileOutput: seq[string]
  for ins in final:
    if ins.kind == HexOrLabelKind.LABEL:
      let label = ins.label
      if not labels.contains(label):
        echo fmt"Undefined label: {label}"
        return
      let labelHex = labels[label]
      fileOutput.add fmt"{(labelHex shr 8):02x}"
      fileOutput.add fmt"{(labelHex and 0xFF):02x}"
    else:
      fileOutput.add fmt"{ins.hex:02x}"

  outputFileStream.writeLine("v2.0 raw")
  outputFileStream.writeLine(join(fileOutput, " "))
  echo fmt"Successfully compiled to: {outputFile}"

proc parse(inputFile: string, outputFile: string) =
  var final: seq[HexOrLabel]
  var labels = initTable[string, int]()
  var offset = 0

  let inputFileStream = newFileStream(inputFile, fmRead)

  if isNil(inputFileStream):
    echo "Input file is empty. Aborting."
    return

  var line: string
  while inputFileStream.readLine(line):
    line = line.replace(re("//.*"), "").strip()
    if line.len == 0:
      continue

    if handleLabels(line, labels, offset):
      continue

    if not translateInstructions(line, offset, final):
      return

  writeOutput(final, labels, outputFile)

var inputFiles: seq[string]
var outputFile: string = ""
for kind, key, val in getopt():
  case kind
  of cmdArgument:
    inputFiles.add(key)
  of cmdLongOption, cmdShortOption:
    case key
    of "o":
      outputFile = val
  of cmdEnd:
    discard

for inputFile in inputFiles:
  echo fmt"Processing {inputFile}"
  if outputFile.strip().len == 0 or inputFiles.len > 1:
    outputFile = changeFileExt(inputFile, "o")
  parse(inputFile, outputFile)
